library/psa_crypto.c

```diff
@@ -2052,13 +2052,21 @@ psa_status_t psa_copy_key( mbedtls_svc_key_id_t source_key,
      * */
     if( psa_key_lifetime_is_external( actual_attributes.core.lifetime ) )
     {
+        /* The actual attributes that we received from the user could have
+         * zero values for key bits and type. It is safe to inherit these
+         * from the source key to get the key buffer size */
+        actual_attributes.core.bits = source_slot->attr.bits;
+        actual_attributes.core.type = source_slot->attr.type;
```

Comment:
For transparent keys:

1. `psa_validate_optional_attributes` checks that the optional attributes (type, size) in `specified_attributes` (the same that have been copied to `actual_attributes`) are either omitted or equal to the source values.
1. `psa_start_key_creation` sets the optional attributes only if they were passed (because it reads `actual_attributes`).
1. `psa_copy_key_material` sets the value of the attributes in the new slot.

For opaque keys, now:

1. `psa_validate_optional_attributes` checks that the optional attributes are omitted or have the correct values.
1. `psa_start_key_creation` sets the optional attributes only if they were passed.
1. The code here sets the values of the optional attributes in `actual_attributes`.
1. The code below sets the values of the optional attributes in the new slot.

I think the code should be simplified to have the same flow in both cases: set the type and size in `actual_attributes` between the call to `psa_validate_optional_attributes` and the call to `psa_start_key_creation`.

With the current flow, since the behavior of filling the key type in size is different for transparent and opaque keys, we'd need extra unit tests for that for opaque keys (duplicating “Copy fail: incorrect type in attributes” and “Copy fail: incorrect size in attributes”). If the flow is the same, having these additional test cases would be nice but is less important.

Reply:
YEs, that's cleaner and would mean we knock off all the source to target attribute write after the copy.
Done !